### ISSUE TITLE
Feature Request - include dropbox.py script and unzip util

### ISSUE THREAD

**Issue by fiveangle** [otherguy/docker-dropbox#16]
Thanks again for maintaining this up-to-date container.  Your latest update is working great !

I'm struggling with tracking down why my dropbox instance stalls at "Downloading 304 files..." forever.  I suspect it may be an incompatibility between some existing files present in the Dropbox cloud vs. supported attributes on the EXT4 file system (like files that came from mac OS trying to pulled down to EXT4, etc).

There is a python script that Dropbox provides to interact with the Dropbox daemon in order to provide additional info since syslog doesn't have verbose details of the files/directories being operated on.  Please consider including this by default in your image, as it is instrumental in troubleshooting when problems occur:

https://help.dropbox.com/installs-integrations/desktop/linux-commands
https://linux.dropbox.com/packages/dropbox.py

(Also, including "unzip" and "vi" (or vim) would be welcome additions too ;)

Thanks for considering !

**Comment by fiveangle** [comment]
found some more interesting tidbits on: https://help.dropbox.com/installs-integrations/sync-uploads/files-not-syncing

> Monitoring more than 10,000 folders on Linux
> The Linux version of the Dropbox desktop app is limited from monitoring more than 10,000 folders by default. Anything more than that is not watched and, therefore, ignored when syncing. There's an easy fix for this. Open a terminal and enter the following:
> 
> echo fs.inotify.max_user_watches=100000 | sudo tee -a /etc/sysctl.conf; sudo sysctl -p
> This command will tell your system to watch up to 100,000 folders. Once the command is entered and you enter your password, Dropbox will immediately resume syncing.

I haven't tested your installation so you have have already addressed, but I thought I'd mention it in case 😃

**Comment by otherguy** [comment]
> There is a python script that Dropbox provides to interact with the Dropbox daemon in order to provide additional info since syslog doesn't have verbose details of the files/directories being operated on. Please consider including this by default in your image, as it is instrumental in troubleshooting when problems occur:
> 
> https://help.dropbox.com/installs-integrations/desktop/linux-commands
> https://linux.dropbox.com/packages/dropbox.py

What is your use case here? From what I can see, this python script does nothing the normal dropbox daemon cannot already do. Have you seen in the `README` that you can run commands like this on the host OS?

    $ docker exec -it dropbox gosu dropbox dropbox status

If you're already inside the container, simply do

    $ gosu dropbox dropbox status

You can run this command to see what's available:

    $ docker exec -it dropbox gosu dropbox dropbox help

> (Also, including "unzip" and "vi" (or vim) would be welcome additions too ;)

Generally containers should be single-purpose. Can you tell me what you are trying to do that requires `unzip` and `vim` _inside_ the container?

> I'm struggling with tracking down why my dropbox instance stalls at "Downloading 304 files..." forever. I suspect it may be an incompatibility between some existing files present in the Dropbox cloud vs. supported attributes on the EXT4 file system (like files that came from mac OS trying to pulled down to EXT4, etc).

It's possible but that is unlikely. How are you starting the container (full command)?



> found some more interesting tidbits on: https://help.dropbox.com/installs-integrations/sync-uploads/files-not-syncing
> 
> > Monitoring more than 10,000 folders on Linux
> > The Linux version of the Dropbox desktop app is limited from monitoring more than 10,000 folders by default. Anything more than that is not watched and, therefore, ignored when syncing. There's an easy fix for this. Open a terminal and enter the following:
> > echo fs.inotify.max_user_watches=100000 | sudo tee -a /etc/sysctl.conf; sudo sysctl -p
> > This command will tell your system to watch up to 100,000 folders. Once the command is entered and you enter your password, Dropbox will immediately resume syncing.
> 
> I haven't tested your installation so you may have already addressed, but I thought I'd mention it in case 😃

Unfortunately this is not something I can put into the container. This needs to be set on your host OS. The container neither has its own filesystem, not does it have `sysctl`.

**Comment by hughker** [comment]
@fiveangle I'm running this on a Synology and I have to run the `echo fs.inotify.max_user_watches=100000 | sudo tee -a /etc/sysctl.conf; sudo sysctl -p` everytime the container starts, then it seems to continue syncing without issues.

For reference, my one account has 877GB and the other had 5.27TB. I haven't tried this with the 5.27TB account, as I'm waiting on a new unit to arrive and configure the process; however, running it with my 877GB account, it does work. When I setup the 5.27TB account, I will monitor the sync process after a couple days and see if I have to update the number even higher.

Thank you @otherguy for putting this together, it's been working great in my beta testing.

**Comment by fiveangle** [comment]
I appoligize for my lack of knowledge on docker container command management (I am more familiar with freebsd jails so should have expected some special treatment was required).  Thank you for the education and your patience.  With your pointing in the right direction, I think I'm now well equipped w/o any changes:

```
[root@0dc127559859:/opt/dropbox/Dropbox# dropbox status
Dropbox isn't running!
root@0dc127559859:/opt/dropbox/Dropbox# gosu dropbox dropbox status
Syncing 304 files
Downloading 304 files...
root@0dc127559859:/opt/dropbox/Dropbox#]

```

The vi request was only in order to review environment files from a terminal since cat and tail weren't sufficient, since we have no knowledge of what the container defaults are, but I have since found that the "more" pager is installed, which is fully sufficient (I didn't know any gnu fileutils based distro continued to have "more" vs it linked to the gnu "less" fileutils, but this pager is plenty sufficent for the rare investigative troubleshooting purposes).

@hughker - I'm guessing your NAS doesn't allow persistently modifying the `/etc/*` contents ?   If it does, try `echo fs.inotify.max_user_watches=100000 >> /etc/sysctl.d/100local.conf` then reload with `sysctl -p` (or reboot) and see if that doesn't make your sync with >10000 folders succeed inside the container w/o having to manually specify every time 🤷‍♂️ 

Thanks again to both of you

**Comment by fiveangle** [comment]
`It's possible but that is unlikely. How are you starting the container (full command)?`

Missed this question before... I'm spinning the container up with docker-compose:

```
root@kobets:/var/lib/docker/recipes/dropbox# cat docker-compose.yml
---
version: "2"
services:
  dropbox:
    image: otherguy/dropbox
    container_name: dropbox
    environment:
      - DBOX_UID=1000
      - DBOX_GID=1000
      - TZ=America/Los_Angeles
    volumes:
      - /var/lib/docker/volumes/dropbox/config/dropbox:/opt/dropbox/.dropbox
    network_mode: 'host'
    restart: unless-stopped
```

I'm suspecting files with " in their name but that's just a hunch.  Perhaps db does code page octal translation like most others do, but who knows.  I'll update for posterity if I find anything interesting.

**Comment by otherguy** [comment]
@hughker @fiveangle to fix the `max_user_watches` issue on Synology NAS devices permanently, go to Control Panel > Task Scheduler. Create a New Task > Triggered Task > User-defined script.

In General Settings tab: 
- give it a name (e.g. “Dropbox sysctl inotify fix”) 
- Set `Event` to `Boot-up`
- Set `User` to `root`
- Check `Enabled`

In the Task Settings tab: 
- enter this in the `Run Command` box: `sh -c '(sleep 60 && echo 204800 > /proc/sys/fs/inotify/max_user_watches)&'`

Reboot your Synology, and verify that this worked by logging in via SSH and issuing this command (you might need to make sure that 60 seconds have passed since boot):

    # sysctl fs.inotify.max_user_watches

Then run the container with the `--privileged` flag.

**Comment by otherguy** [comment]
> I appoligize for my lack of knowledge on docker container command management (I am more familiar with freebsd jails so should have expected some special treatment was required). 

No need to apologize. I know next to nothing about FreeBSD jails ;)
Please always feel free to open an issue here if you need support or help with this container image.
 
> ```
> [root@0dc127559859:/opt/dropbox/Dropbox# dropbox status
> Dropbox isn't running!
> root@0dc127559859:/opt/dropbox/Dropbox# gosu dropbox dropbox status
> Syncing 304 files
> Downloading 304 files...
> ```

This is expected. If you run `dropbox status`, it's running as `root` but the dropbox daemon is using its own user account. That's why you get `Dropbox isn't running!`.

`gosu dropbox` is essentially the same as `sudo -u dropbox` [but for Docker containers](https://github.com/tianon/gosu).

> The vi request was only in order to review environment files from a terminal since cat and tail weren't sufficient. We have no knowledge of what the container defaults are so it's nice to be able to review with something, but I have since found that the "more" pager is installed, which is fully sufficient (I didn't know any gnu fileutils based distro continued to have "more" vs it linked to the gnu "less" fileutils, but this pager is plenty sufficent for the rare investigative troubleshooting purposes).

It's rarely necessary to actually `exec` into the container. There are multiple ways you can view files in the container:

* Copy the file to your host os: `docker cp dropbox:/etc/sysctl.conf sysctl.conf` (the reverse is also possible)
* When starting the container, map `/etc` (or whatever directory you're interested in) to a local directory, using `--volume
* Locally, [Visual Studio Code](https://code.visualstudio.com) lets you bind a whole project to a container and edit every file
* ❗️Or, my favorite version, use [this little script](https://gist.github.com/dmohs/b50ea4302b62ebfc4f308a20d3de4213) to edit any file inside a Docker container using your `$EDITOR`

> @hughker - I'm guessing your NAS doesn't allow persistently modifying the `/etc/*` contents ? If it does, try `echo fs.inotify.max_user_watches=100000 >> /etc/sysctl.d/100local.conf` then reload with `sysctl -p` (or reboot) and see if that doesn't make your sync with >10000 folders succeed inside the container w/o having to manually specify every time 🤷‍♂️

See my reply above for a solution that survives Synology OS updates 🤫

> ```
> root@kobets:/var/lib/docker/recipes/dropbox# cat docker-compose.yml
> ---
> version: "2"
> services:
>   dropbox:
>     image: otherguy/dropbox
>     container_name: dropbox
>     environment:
>       - DBOX_UID=1000
>       - DBOX_GID=1000
>       - TZ=America/Los_Angeles
>     volumes:
>       - /var/lib/docker/volumes/dropbox/config/dropbox:/opt/dropbox/.dropbox
>     network_mode: 'host'
>     restart: unless-stopped
> ```

Please verify that `UID=1000` and `GID=1000` is _actually_ what you want! This should match `id -u` and `id -g` of the user starting the container, or you will run into permission (and syncing) issues.

[Since you're on Linux, I would also recommend](https://github.com/otherguy/docker-dropbox#time-zones) mounting `/etc/timezone` and `/etc/localtime` into the container instead of setting `TZ` manually. Saves you from updating configuration down the road.

You're also not mounting `/opt/dropbox/Dropbox` (the actual files synced by Dropbox), so I'm certain it's not an EXT4 issue.

**Comment by hughker** [comment]
@otherguy wow, that worked perfectly, thank you! I did have to restart the Dropbox container just so that it could see the new value, but I'm sure I could even make another script to just restart the container as well.

---

*I can for sure create a new issue if you'd like, I just was trying to save the issues from being flooded, since we already had talked about Synology here.*

---

On a related Synology question, when I try to start the container via CLI as my user `winston` it doesn't work, so I have to `sudo su` and then run the command again and it'll work just fine going forward without any further errors:

```shell
winston@synology-vm-01:/$ docker run --privileged --detach -it --restart=always --name=dropbox --net="host" -e "TZ=America/Los_Angeles" -e "DROPBOX_UID=0" -e "DROPBOX_GID=100" -v "/volume1/docker/dropbox-settings:/opt/dropbox/.dropbox" -v "/volume1/dropbox:/opt/dropbox/Dropbox" otherguy/dropbox:latest
docker: Got permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Post http://%2Fvar%2Frun%2Fdocker.sock/v1.39/containers/create?name=dropbox: dial unix /var/run/docker.sock: connect: permission denied.
See 'docker run --help'.
winston@synology-vm-01:/$ sudo su
Password: 
ash-4.3# docker run --privileged --detach -it --restart=always --name=dropbox --net="host" -e "TZ=America/Los_Angeles" -e "DROPBOX_UID=0" -e "DROPBOX_GID=100" -v "/volume1/docker/dropbox-settings:/opt/dropbox/.dropbox" -v "/volume1/dropbox:/opt/dropbox/Dropbox" otherguy/dropbox:latest
4e9e1637219da87333eee08d5601b9ffe9b81a1f864a85b029199a3a0a5cc589
ash-4.3# 
```

I'm using `DROPBOX_GID=100` which is my user `winston` GID:

```shell
winston@synology-vm-01:/$ id winston
uid=1026(winston) gid=100(users) groups=100(users),101(administrators)
```

And, `DROPBOX_UID=0` which is `root` UID:

```shell
winston@synology-vm-01:/$ id root
uid=0(root) gid=0(root) groups=0(root),2(daemon),19(log)
```

Let me know if you know of a better way to do this, or if you think this is fine for what is needed, I am open to suggestions.

**Comment by hughker** [comment]
@fiveangle I see what you're talking about with your original comment of it "stalling" now that the majority of my files have synced. 

I'm getting the following on my system:

```shell
Syncing 17 files
Downloading 17 files...
...
Syncing 17 files • 58 mins
Downloading 17 files (10.1 KB/sec, 58 mins)
Syncing 17 files • 47 mins
Downloading 17 files (12.4 KB/sec, 47 mins)
Syncing 17 files
Downloading 17 files...
```

I've reached out to Dropbox Support on an existing Ticket I have open regarding Dropbox Linux CLI and will report back if I learn anything. I specifically asked if there was a way to view which files are having issues syncing, so they can be fixed. 

As far as I can tell: this isn't an issue with @otherguy container.

**Comment by otherguy** [comment]
> @otherguy wow, that worked perfectly, thank you! I did have to restart the Dropbox container just so that it could see the new value, but I'm sure I could even make another script to just restart the container as well.

This should not be necessary. The container doesn't have its own filesystem if you are mounting `/opt/dropbox/Dropbox` to a local folder. Therefore the container doesn't need to pick up that value.

If your Synology is booting that fast, you can also lower the 60 second delay in the command. 5 seconds should be fine too. Or get rid of the delay altogether.

> On a related Synology question, when I try to start the container via CLI as my user `winston` it doesn't work, so I have to `sudo su` and then run the command again and it'll work just fine going forward without any further errors:

That makes sense. If you want to run Docker containers as your user `winston` , follow these steps:

- Create the `docker` group from the Synology UI or CLI (`sudo synogroup --add docker`)
- Change group of the `docker.sock` (`sudo chown root:docker /var/run/docker.sock`)
- Assign your user to the `docker` group in the Synology UI or CLI (`sudo synogroup --member docker {username}`)
- Log out from SSH and back in with your user

Now you can interact with the Docker daemon normally.

> `winston@synology-vm-01:/$ sudo su`

On a side note, you probably want to use `sudo su -` or `sudo -i` instead of just `sudo su`, so you get a normal login shell.
 
> I'm using `DROPBOX_GID=100` which is my user `winston` GID:
> 
> ```shell
> winston@synology-vm-01:/$ id winston
> uid=1026(winston) gid=100(users) groups=100(users),101(administrators)
> ```
> 
> And, `DROPBOX_UID=0` which is `root` UID:
>
> Let me know if you know of a better way to do this, or if you think this is fine for what is needed, I am open to suggestions.

You can, if that's what you want. Files created by dropbox will be owned by `root:users` then (because `uid=0=root` and `gid=100=users`). Depending on your umask, you might run into issues when your `winston` user is editing files.

I don't know what you're trying to achieve but generally you could set `DROPBOX_UID= 1026` and `DROPBOX_GID= 100` to have files owned by `winston:users`.


> I'm getting the following on my system:
> 
> ```shell
> Syncing 17 files
> Downloading 17 files...
> ...
> Syncing 17 files • 58 mins
> Downloading 17 files (10.1 KB/sec, 58 mins)
> Syncing 17 files • 47 mins
> Downloading 17 files (12.4 KB/sec, 47 mins)
> Syncing 17 files
> Downloading 17 files...
> ```
> 
> I've reached out to Dropbox Support on an existing Ticket I have open regarding Dropbox Linux CLI and will report back if I learn anything. I specifically asked if there was a way to view which files are having issues syncing, so they can be fixed.
> 
> As far as I can tell: this isn't an issue with @otherguy container.

It probably isn't an issue with the container. How long have you let it run? And did it ever finish by itself?

**Comment by fiveangle** [comment]
@hughker yes, with the info on running the dropbox command in the container context above from @otherguy I was surprised that after several hours of iterating over every file, there were none still synching (!)

That's when I switched it up and found that it's in fact stuck directories that can't pull down from the server to the local system:

find ./Dropbox  -type d -exec gosu dropbox dropbox filestatus {} \; | grep 'syncing'

This uncovered directories that had no files in them with status "syncing" yet were themselves marked as such. Most of them seemed to be located only on the Dropbox mothership.  I ended up moving any that were on my client outside of the db folder, allowed syncing of any files to complete (status "stalled" to unchanging), then manually deleted the folder on db web interface, watched status in the container stall again at a lower number, then moved the directory back.  Wash, rinse, repeat, and status is now "Up to date".

And people wonder why I've just switched to Nextcloud 😛 (and hence this effort with dropbox container).  I have a remaining small subset of files I will sync to dropbox with the free (3-client limit) tier and that's what this container is for... I just mount the "dropbox" subfolder that I've created inside my nextcloud volume so this container gets me that off-site backup for mission critical stuff + access for all the dropbox-integrated mobile apps, while not being shackled to syncing only 3 clients.

Thanks again Alex for all your great effort !

**Comment by hughker** [comment]
> If your Synology is booting that fast, you can also lower the 60 second delay in the command. 5 seconds should be fine too. Or get rid of the delay altogether.

@otherguy until I get the physical unit, I'll probably wait to change the delay or remove it. But for now, I just created a 2nd task `sh -c '(sleep 120 && docker restart dropbox)'` that runs after the 1st task you suggested runs and everything works perfectly for now. 

---

> That makes sense. If you want to run Docker containers as your user `winston` , follow these steps:
> 
> * Create the `docker` group from the Synology UI or CLI (`sudo synogroup --add docker`)
> * Change group of the `docker.sock` (`sudo chown root:docker /var/run/docker.sock`)
> * Assign your user to the `docker` group in the Synology UI or CLI (`sudo synogroup --member docker {username}`)
> * Log out from SSH and back in with your user
> 
> Now you can interact with the Docker daemon normally.
> 
> >

@otherguy that is exactly what I was looking for, I just couldn't find the Synology specific commands to make it happen! As expected, now `docker ps` works as intended from my shell.

---

> I don't know what you're trying to achieve but generally you could set `DROPBOX_UID= 1026` and `DROPBOX_GID= 100` to have files owned by `winston:users`.

@otherguy when I changed it to your suggestion, I got the following error when the container was starting up:

```shell
!! dropbox: failed to create log file (Permission denied)!
```

But when I put it back to `DROPBOX_UID= 0` then it didn't have that error about the log file. If the log file really isn't needed, then I'm okay with using `1026` I just saw that error awhile back and then changed it to `0` so that message would go away.

---

> It probably isn't an issue with the container. How long have you let it run? And did it ever finish by itself?

@fiveangle thanks for that insight. I ran `docker exec -it dropbox gosu dropbox dropbox filestatus | grep 'syncing'` and then excluded whatever directories were listed and now it's returning `Up to date`. Then I added an example trouble directory back to sync, it then returned back the same loop as we experienced before.

So I took your advice, moved the trouble directory from `/dropbox` to `/holding` then deleted it from Dropbox.com, waited for it to be `Up to date` then moved it from `/holding` back to `/dropbox` and then waited for it to return `Up to date` again. 

Certainly a bit of a pain to do that, but at least we now have a work-around to get it all to be `Up to date`!

---

Thank you again for your continued assistance @otherguy and thank you @fiveangle for that insight on how to find what was stuck "syncing".

**Comment by otherguy** [comment]
> And people wonder why I've just switched to Nextcloud 😛 (and hence this effort with dropbox container). I have a remaining small subset of files I will sync to dropbox with the free (3-client limit) tier and that's what this container is for...

@fiveangle Same here :) With their ridiculous limits and pricing, I switched away a long time ago. But it's still useful for some things and apps that only sync to Dropbox.

> @otherguy until I get the physical unit, I'll probably wait to change the delay or remove it. But for now, I just created a 2nd task `sh -c '(sleep 120 && docker restart dropbox)'` that runs after the 1st task you suggested runs and everything works perfectly for now.

@hughker you can still lower it to 5 or 10 seconds or even remove it. I just have a ton of things starting when my Synology boots which is why I came up with a delay.

But thinking about it, it actually makes no sense. As long as the kernel is booted, you can set that value. And it should actually happen before any other programs start.

That said, you do not need to restart the Docker container after setting this!

> @otherguy when I changed it to your suggestion, I got the following error when the container was starting up:
> 
> ```shell
> !! dropbox: failed to create log file (Permission denied)!

This does not matter as you're only interacting with the `stdout` logs anyway (`docker logs`). If you run it as `root`, the daemon will try to log to `/tmp/dropbox-antifreeze-{random-id}`.

This will, over time, fill up your hard drive. So this "error" is actually a good thing!

**Comment by hughker** [comment]
> @hughker you can still lower it to 5 or 10 seconds or even remove it. I just have a ton of things starting when my Synology boots which is why I came up with a delay.
> 
> But thinking about it, it actually makes no sense. As long as the kernel is booted, you can set that value. And it should actually happen before any other programs start.
> 
> That said, you do not need to restart the Docker container after setting this!
> 
> >

@otherguy sounds good, thank you for that insight - I will remove the delay and Docker restart once I get the physical unit. 

---

> This does not matter as you're only interacting with the `stdout` logs anyway (`docker logs`). If you run it as `root`, the daemon will try to log to `/tmp/dropbox-antifreeze-{random-id}`.
> 
> This will, over time, fill up your hard drive. So this "error" is actually a good thing!

@otherguy that's great to know! I will put it back to the non-root UID once I get the physical unit to prevent the logs from filling up the drive.
